### ISSUE TITLE
Shadowing parameters are sometimes renamed

### ISSUE THREAD

**Issue by nad** [agda/agda#2018]
Agda rejects the following code:

``` agda
module _ {A : Set} where

record R {A : Set} : Set where

postulate
  B : Set
  r : R {A = B}
```

Error message:

```
Function does not accept argument {A = _}
when checking that {A = B} is a valid argument to a function of
type {A₁ : Set} → Set
```

I don't think that Agda should change the type of `R` in this way.

**Comment by andreasabel** [comment]
I wonder if that ever worked or which commit made it fail...

**Comment by nad** [comment]
With Agda 2.3.2.2:

```
Set should be a function type, but it isn't
when checking that {A = B} are valid arguments to a function of
type Set
```

With Agda 2.4.0.2:

```
Function does not accept argument {A = _}
when checking that {A = B} are valid arguments to a function of
type {A₁ : Set} → Set
```

**Comment by andreasabel** [comment]
Both point to the fact that the named argument {A = B} was not accepted, only the error message has changed (I remember improving the error messages at some point).  So either the problem here was introduced earlier, or this was never handled correctly.

**Comment by nad** [comment]
I suspect that this code has never been accepted.

If you replace `R {A = B}` with `R {A₁ = B}`, then Agda 2.3.2.2 still rejects the code, but Agda 2.4.0.2 accepts it. Thus the situation was made worse between 2.3.2.2 and 2.4.0.2.

**Comment by andreasabel** [comment]
While researching this issue, I found the following minor internal inconsistency in the record field telescope:

``` agda
{-# OPTIONS -v tc.term.rec:20 #-}

module _ where

module _ (A : Set) where

  record R : Set where
    constructor c
    field f : A

test : ∀{A} → A → R A
test a = record { f = a }

-- Debug printing:

-- checking record expression record { f = a }
--   r   = RecordConstructor._.R
--   xs  = [f]
--   ftel= (A : Set) (f : A)
--   con = RecordConstructor._.R.c(inductive)[RecordConstructor._.R.f]
-- finished record expression
```

Note that the parameter in `ftel` does not have the correct hiding.`

**Comment by andreasabel** [comment]
We have the same problem for the fields:

``` agda
module _ {A : Set} where

record R {A : Set} : Set where
  constructor rc
  field f : A

test-f : ∀ B → R {A = B} → B
test-f B = R.f {A = B}
```

reports

```
Function does not accept argument {A = _}
when checking that {A = B} is a valid argument to a function of
type {A₁ : Set} → R → A₁
```

**Comment by andreasabel** [comment]
This problem is endemic, we have it also with modules:

``` agda
module _ {A : Set} where

postulate B : Set

module M {A : Set} where

open M {A = B}
```

reports

```
The arguments to M do not fit the telescope {A₁ : Set}
when checking the module application module _ = M {A = B}
```

There is a principal design flaw, which is a combination of these two mechanisms:
1. Declarations get automatically lifted to the top-level via `addConstant` and `addSection`, abstracted over the current context.
2. When new bindings are added to the context, they are renamed to not conflict with previous bindings.

Imho, 2. is principally wrong.  Instead the printer should mark shadowed names as such (this is, e.g., what Twelf does).
However, changing 2. might be a nightmare, we will have to radically change the printer. @UlfNorell

**Comment by andreasabel** [comment]
Horray for fixing this!

**Comment by andreasabel** [comment]
@UlfNorell: You have not answered my question from 14 Jan 2017 in https://github.com/agda/agda/commit/67d50a9159147f0aa990aef4ae3ccbbfdd2b56df

**Comment by UlfNorell** [comment]
> @UlfNorell: Why is underAbstraction not implemented in terms of addContext, which also does the unshadow/fresh magic?
Why does underAbstraction the realName thing and addContext not?

I don't know.

**Comment by andreasabel** [comment]
The purpose of my question is that my question is answered in a comment in the code, not here on the issue tracker.

Don't discourage your code reviewers!

**Comment by andreasabel** [comment]
Looking at the history of test/interaction/Issue231.out , it is likely that regression #2729 was introduced by the fix of this issue: https://github.com/agda/agda/commit/67d50a9159147f0aa990aef4ae3ccbbfdd2b56df

Probably the fix of #2424 also contributes: https://github.com/agda/agda/commit/e60c544ed9f2234c7a2021adfae356811025d4bc

**Comment by andreasabel** [comment]
I will push a fix for #2729 which reverts my fix for #2424 and refines @UlfNorell 's fix introduced for this issue.
